device/cluster.cpp

```diff
@@ -1403,43 +1387,68 @@ void Cluster::set_fallback_tlb_ordering_mode(const std::string& fallback_tlb, ui
+// TODO: this is in the wrong place, it should be in the TTDevice.
+// It should also happen at the same time the huge pages or sysmem buffers are
+// allocated/pinned/mapped.
 void Cluster::init_pcie_iatus() {
     int num_enabled_devices = m_tt_device_map.size();
     log_debug(LogSiliconDriver, "Cluster::init_pcie_iatus() num_enabled_devices: {}", num_enabled_devices);
 
-    for (auto& src_device_it : m_tt_device_map) {
-        int logical_id = src_device_it.first;
-        PCIDevice* src_pci_device = src_device_it.second->get_pci_device();
+    for (auto& [logical_id, tt_device] : m_tt_device_map) {
+        PCIDevice* pci_device = tt_device->get_pci_device();
 
         // TODO: with the IOMMU case, I think we can get away with using just
         // one iATU region for WH.  (On BH, we don't need iATU).  We can only
         // cover slightly less than 4GB with WH, and the iATU can cover 4GB.
         // Splitting it into multiple regions is fine, but it's not necessary.
         //
-        // ... something to consider when this code is refactored into PCIDevice
-        // where it belongs.
-
-        // Device to Host (multiple channels)
-        for (int channel_id = 0; channel_id < src_pci_device->get_num_host_mem_channels(); channel_id++) {
-            hugepage_mapping hugepage_map = src_pci_device->get_hugepage_mapping(channel_id);
-            if (hugepage_map.mapping) {
-                std::uint32_t region_size = hugepage_map.mapping_size;
-                if (channel_id == 3) {
+        // Update: unfortunately this turned out to be unrealistic.  For the
+        // IOMMU case, the easiest thing to do is fake that we have hugepages
+        // so we can support the hugepage-inspired API that the user application
+        // has come to rely on.  In that scenario, it's simpler to treat such
+        // fake hugepages the same way we treat real ones -- even if underneath
+        // there is only a single buffer.  Simple is good.
+        //
+        // With respect to BH: it turns out that Metal has hard-coded NOC
+        // addressing assumptions for sysmem access.  First step to fix this is
+        // have Metal ask us where sysmem is at runtime, and use that value in
+        // on-device code.  Until then, we're stuck programming iATU.  A more
+        // forward-looking solution is to abandon the sysmem API entirely, and
+        // have the application assume a more active role in managing the memory
+        // shared between host and device.  UMD would be relegated to assisting
+        // the application set up and tear down the mappings.  This is probably
+        // a unrealistic for GS/WH, but it's a good goal for BH.
+        //
+        // Until then...
+        //
+        // For every 1GB channel of memory mapped for DMA, program an iATU
+        // region to map it to the underlying buffer's IOVA (IOMMU case) or PA
+        // (non-IOMMU case).
```

Comment:
I see you've refactored this iatu code a bit, and thanks, it is going in the right direction! But I'm missing what here is iommu specific? Where exactly does the code path split for IOMMU case compared to hugepages? I see 
`For the IOMMU case, the easiest thing to do is fake that we have hugepages`

Or was this split already done in previous PR?

Reply:
Fake hugepages were introduced in the previous PR.  There's nothing here that is IOMMU-specific, although what happens here is necessary for the IOMMU case due to Metal's reliance on hard-coded NOC addresses.  That is, we need to program iATU anyway.  The difference between IOMMU case vs no IOMMU+hugepage case here is the contents of `hugepage_map`.  For IOMMU case, the `physical_address` is not actually a physical address -- it's an I/O virtual address.